File: src/lib/forum-actions.ts
"use server";

/**
 * Team forum writes.
 *
 * **Not a security boundary.** The only real check below is `assertMembership`
 * — it confirms the claimed author is on the team the token unlocks, nothing
 * more. Anyone holding a team's link can already see every member's name, so
 * they can post, edit, or delete as any one of them; "remember my name" in the
 * browser is a convenience for an honest student, not an identity check. Do
 * not read the `authorStudentId`/ownership checks here as authorization —
 * they stop a UI accident, not a determined teammate. Real accountability is
 * the audit log (actor name + IP on every write) and `regenerateTeamLink` to
 * rotate a leaked link.
 */

import { and, eq, gt, sql } from "drizzle-orm";
import { notFound, redirect } from "next/navigation";

import { db } from "@/db";
import { forumMessages } from "@/db/schema";
import { recordAudit } from "./auth";
import { assertMembership, getTeamByToken } from "./team-access";

const MAX_BODY_LENGTH = 4000;
const DUPLICATE_WINDOW_MS = 60_000;
const MAX_MESSAGES_PER_TEAM = 500;

const str = (form: FormData, key: string) => String(form.get(key) ?? "").trim();
const num = (form: FormData, key: string) => Number(form.get(key));

const back = (token: string, message: string) =>
  redirect(`/t/${token}?forumError=${encodeURIComponent(message)}#forum`);

export async function postMessage(formData: FormData): Promise<void> {
  const token = str(formData, "token");
  const ctx = await getTeamByToken(token);
  if (!ctx) notFound();

  const studentId = num(formData, "authorStudentId");
  const member = ctx.members.find((m) => m.id === studentId);
  if (!member || !(await assertMembership(ctx.team.id, studentId))) {
    back(token, "Pick your name from the list before posting.");
  }

  const body = str(formData, "body").slice(0, MAX_BODY_LENGTH);
  if (!body) back(token, "Write something before sending.");

  // A reply to a reply is flattened onto the top-level message it replies to
  // — one level deep is all the UI needs, and it keeps every query a single
  // pass with no recursion.
  const rawParentId = num(formData, "parentId");
  let parentId: number | null = null;
  if (Number.isInteger(rawParentId)) {
    const parent = await db.query.forumMessages.findFirst({
      where: and(
        eq(forumMessages.id, rawParentId),
        eq(forumMessages.teamId, ctx.team.id),
      ),
    });
    if (parent) parentId = parent.parentId ?? parent.id;
  }

  const recentDuplicate = await db.query.forumMessages.findFirst({
    where: and(
      eq(forumMessages.teamId, ctx.team.id),
      eq(forumMessages.authorStudentId, studentId),
      eq(forumMessages.body, body),
      gt(forumMessages.createdAt, Date.now() - DUPLICATE_WINDOW_MS),
    ),
  });
  if (recentDuplicate) {
    back(token, "That looks like the message you just sent.");
  }

  const countRow = await db
    .select({ n: sql<number>`count(*)` })
    .from(forumMessages)
    .where(eq(forumMessages.teamId, ctx.team.id))
    .get();
  if (Number(countRow?.n ?? 0) >= MAX_MESSAGES_PER_TEAM) {
    back(
      token,
      "This team's board is full. Ask your course responsible for help.",
    );
  }

  await db.insert(forumMessages).values({
    teamId: ctx.team.id,
    parentId,
    authorStudentId: studentId,
    authorName: member!.name,
    body,
  });

  await recordAudit({
    action: "forum.posted",
    actorName: member!.name,
    teamId: ctx.team.id,
    detail: parentId ? "reply" : "new thread",
  });

  redirect(`/t/${token}#forum`);
}

export async function editMessage(formData: FormData): Promise<void> {
  const token = str(formData, "token");
  const ctx = await getTeamByToken(token);
  if (!ctx) notFound();

  const messageId = num(formData, "messageId");
  const message = await db.query.forumMessages.findFirst({
    where: and(eq(forumMessages.id, messageId), eq(forumMessages.teamId, ctx.team.id)),
  });
  if (!message || message.deletedAt) notFound();

  const body = str(formData, "body").slice(0, MAX_BODY_LENGTH);
  if (!body) back(token, "A message can't be edited down to nothing.");

  await db
    .update(forumMessages)
    .set({ body, editedAt: Date.now() })
    .where(eq(forumMessages.id, messageId));

  await recordAudit({
    action: "forum.edited",
    actorName: message.authorName,
    teamId: ctx.team.id,
  });

  redirect(`/t/${token}#forum`);
}

export async function deleteMessage(formData: FormData): Promise<void> {
  const token = str(formData, "token");
  const ctx = await getTeamByToken(token);
  if (!ctx) notFound();

  const messageId = num(formData, "messageId");
  const message = await db.query.forumMessages.findFirst({
    where: and(eq(forumMessages.id, messageId), eq(forumMessages.teamId, ctx.team.id)),
  });
  if (!message || message.deletedAt) notFound();

  await db
    .update(forumMessages)
    .set({ deletedAt: Date.now() })
    .where(eq(forumMessages.id, messageId));

  await recordAudit({
    action: "forum.deleted",
    actorName: message.authorName,
    teamId: ctx.team.id,
  });

  redirect(`/t/${token}#forum`);
}

File: app/lib/forum_actions.py
"""Team forum writes.

**Not a security boundary.** The only real check below is `assert_membership`
-- it confirms the claimed author is on the team the token unlocks, nothing
more. Anyone holding a team's link can already see every member's name, so
they can post, edit, or delete as any one of them; "remember my name" in the
browser is a convenience for an honest student, not an identity check. Do
not read the membership check here as authorization -- it stops a UI
accident, not a determined teammate. Real accountability is the audit log
(actor name + IP on every write) and regenerating a team's link to rotate a
leaked one.
"""

from __future__ import annotations

import time
from typing import Optional

from app.db import db_lock, get_db
from app.lib.auth import record_audit
from app.lib.team_access import TeamContext, assert_membership

MAX_BODY_LENGTH = 4000
DUPLICATE_WINDOW_MS = 60_000
MAX_MESSAGES_PER_TEAM = 500


def post_message(
    ctx: TeamContext, author_student_id: int, body: str, parent_id: Optional[int], ip: Optional[str]
) -> Optional[str]:
    """Returns an error string, or None on success."""
    member = next((m for m in ctx.members if m["id"] == author_student_id), None)
    if not member or not assert_membership(ctx.team["id"], author_student_id):
        return "Pick your name from the list before posting."

    body = body.strip()[:MAX_BODY_LENGTH]
    if not body:
        return "Write something before sending."

    conn = get_db()

    # A reply to a reply is flattened onto the top-level message it replies
    # to -- one level deep is all the UI needs, and it keeps every query a
    # single pass with no recursion.
    resolved_parent_id = None
    if parent_id is not None:
        parent = conn.execute(
            "SELECT * FROM forum_messages WHERE id = ? AND team_id = ?",
            (parent_id, ctx.team["id"]),
        ).fetchone()
        if parent:
            resolved_parent_id = parent["parent_id"] or parent["id"]

    recent_duplicate = conn.execute(
        "SELECT 1 FROM forum_messages WHERE team_id = ? AND author_student_id = ? "
        "AND body = ? AND created_at > ?",
        (ctx.team["id"], author_student_id, body, int(time.time() * 1000) - DUPLICATE_WINDOW_MS),
    ).fetchone()
    if recent_duplicate:
        return "That looks like the message you just sent."

    count = conn.execute(
        "SELECT count(*) AS n FROM forum_messages WHERE team_id = ?", (ctx.team["id"],)
    ).fetchone()["n"]
    if count >= MAX_MESSAGES_PER_TEAM:
        return "This team's board is full. Ask your course responsible for help."

    with db_lock() as conn:
        conn.execute(
            "INSERT INTO forum_messages (team_id, parent_id, author_student_id, author_name, body) "
            "VALUES (?, ?, ?, ?, ?)",
            (ctx.team["id"], resolved_parent_id, author_student_id, member["name"], body),
        )
        conn.commit()

    record_audit(
        action="forum.posted",
        actor_name=member["name"],
        team_id=ctx.team["id"],
        # The comment itself, not just "reply"/"new thread" -- so an admin
        # reading the notifications list sees what was actually said without
        # having to open the team's forum separately.
        detail=f"{'Reply' if resolved_parent_id else 'New thread'}: {body}",
        ip=ip,
    )
    return None


def edit_message(ctx: TeamContext, message_id: int, body: str) -> tuple[bool, Optional[str]]:
    """Returns (found, error)."""
    conn = get_db()
    message = conn.execute(
        "SELECT * FROM forum_messages WHERE id = ? AND team_id = ?",
        (message_id, ctx.team["id"]),
    ).fetchone()
    if not message or message["deleted_at"]:
        return False, None

    body = body.strip()[:MAX_BODY_LENGTH]
    if not body:
        return True, "A message can't be edited down to nothing."

    with db_lock() as conn:
        conn.execute(
            "UPDATE forum_messages SET body = ?, edited_at = ? WHERE id = ?",
            (body, int(time.time() * 1000), message_id),
        )
        conn.commit()

    record_audit(action="forum.edited", actor_name=message["author_name"], team_id=ctx.team["id"])
    return True, None


def delete_message(ctx: TeamContext, message_id: int) -> bool:
    """Returns whether the message existed."""
    conn = get_db()
    message = conn.execute(
        "SELECT * FROM forum_messages WHERE id = ? AND team_id = ?",
        (message_id, ctx.team["id"]),
    ).fetchone()
    if not message or message["deleted_at"]:
        return False

    with db_lock() as conn:
        conn.execute(
            "UPDATE forum_messages SET deleted_at = ? WHERE id = ?",
            (int(time.time() * 1000), message_id),
        )
        conn.commit()

    record_audit(action="forum.deleted", actor_name=message["author_name"], team_id=ctx.team["id"])
    return True
